infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationAckReader.java

```diff
@@ -111,43 +122,169 @@ private long getMaxReplicatedStreamsTail() {
+
+            // If during snapshot sync nothing has been ack'ed, all replication is remaining
+            if (ackedTimestamp == Address.NON_ADDRESS) {
+                ackedTimestamp = 0;
+            }
+
+            // In Snapshot Sync
+            // Simply subtract the ack'ed timestamp from the global log tail from the time the snapshot sync started.
+            // Note that this is not accurate because the global log tail does not accurately represent the remaining entries
+            // for replicated streams.
+            // When snapshot sync is ongoing, there may be delta updates also. Add those new entries by querying the address maps
+            return ((baseSnapshotTimestamp - ackedTimestamp) +
+                    getTxStreamTotalEntries(baseSnapshotTimestamp, txStreamTail));
         }
 
-        // When in LogEntry Sync, no CP and trim has taken place so the remaining entries can be queried using the
-        // global tail and address maps of the replicated streams
-        if (lastSyncType == LogReplicationMetadata.ReplicationStatusVal.SyncType.LOG_ENTRY) {
-            return calculateRemainingEntriesForLogEntrySync(maxReplicatedStreamTail, ackedTimestamp);
+        // In Log Entry Sync
+        return calculateRemainingEntriesIncrementalUpdates(ackedTimestamp, txStreamTail, currentTxStreamProcessedTs);
+    }
+
+    private long getTxStreamTail(Map<UUID, Long> tailMap) {
+        if (tailMap.containsKey(TRANSACTION_STREAM_ID)) {
+            return tailMap.get(TRANSACTION_STREAM_ID);
         }
 
-        // In Snapshot Sync
-        // Simply subtract the ackedTimestamp from the global log tail from the time the snapshot sync started.
-        // Note that this is not accurate because the global log tail does not accurately represent the remaining entries
-        // for replicated streams.
-        // When snapshot sync is ongoing, there may be delta updates also.  Add those new entries by querying the address maps
-        return ((baseSnapshotTimestamp - ackedTimestamp) +
-            calculateRemainingEntriesForLogEntrySync(maxReplicatedStreamTail, baseSnapshotTimestamp));
+        log.warn("Tx Stream tail not present in sequencer, id={}", TRANSACTION_STREAM_ID);
+        return Address.NON_ADDRESS;
     }
 
-    private long calculateRemainingEntriesForLogEntrySync(long start, long end) {
-        long remainingEntriesToSend = 0;
-        for (String stream : config.getStreamsToReplicate()) {
-            UUID streamId = CorfuRuntime.getStreamID(stream);
-            StreamAddressRange range = new StreamAddressRange(streamId, start, end);
-            StreamAddressSpace addressSpace = runtime.getSequencerView().getStreamAddressSpace(range);
-            remainingEntriesToSend += addressSpace.getAddressMap().getLongCardinality();
+    /**
+     * The calculation of remaining entries during log entry sync takes into account:
+     * - lastAckedTimestamp
+     * - txStreamTail
+     * - currentTxStreamProcessedTs
```

Comment:
Can we add a small comment explaining the role of this param?  It can be a bit confusing..

Reply:
sure, added.